packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -153,32 +153,48 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
   const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
   const mounted = React.useRef(false);
 
+  // Record to store all extra information (e.g. original indexes), where the key is the id of the option
+  const optionsExtraInfoRef = React.useRef<
+    Record<string, { __originalIndex: number }>
+  >({});
+
+  // Function to initialize extra info
+  const initializeExtraInfo = React.useCallback(() => {
+    optionsProp.current.forEach((option, index) => {
+      optionsExtraInfoRef.current[getOptionId(option, id)] = {
+        __originalIndex: index,
+      };
+    });
+  }, [id]);
+
+  // Latest value of the valueProp prop
+  const localValueProp = React.useRef(valueProp);
+  React.useEffect(() => {
+    localValueProp.current = valueProp;
+  }, [valueProp]);
+
   // Latest value of the onChange prop
   const onChangeProp = React.useRef(onChange);
   React.useEffect(() => {
     onChangeProp.current = onChange;
   }, [onChange]);
 
-  // Record to store all extra information (e.g. original indexes), where the key is the id of the option
-  const optionsExtraInfoRef = React.useRef<
-    Record<string, { __originalIndex: number }>
-  >({});
-
-  // Clear the extra info when the options change so that it can be reinitialized below
+  // Latest value of the options prop
+  const optionsProp = React.useRef(options);
```

Comment:
Rename to `optionsRef` since there is `valueProp` and means completely different thing.

Reply:
renamed